packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+
+      // Typecast: Immer WritableDraft incompatibility
+      draftState.subjects[origin].permissions = castDraft(permissions);
+    });
+  }
+
+  /**
+   * Constructs the requested caveats for the permission of the specified
+   * subject origin and target name.
+   *
+   * Throws an error if validation fails.
+   *
+   * @see {@link PermissionController.constructCaveat} for details.
+   * @param origin - The origin of the permission subject.
+   * @param target - The permission target name.
+   * @param requestedCaveats - The requested caveats to construct.
+   * @returns The constructed caveats.
+   */
+  private constructCaveats(
+    origin: OriginString,
+    target: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    requestedCaveats?: unknown[] | null,
+  ): NonEmptyArray<ExtractCaveats<ControllerCaveatSpecification>> | undefined {
+    const caveatArray = requestedCaveats?.map((requestedCaveat) =>
+      this.constructCaveat(origin, target, requestedCaveat),
+    );
+    return caveatArray && isNonEmptyArray(caveatArray)
+      ? caveatArray
+      : undefined;
+  }
+
+  /**
+   * Constructs a caveat for the permission of the specified subject origin and
+   * target name, per the requested caveat argument. This methods validates
+   * everything about the requested caveat, except that its `value` property
+   * is JSON-compatible. It also ensures that a caveat specification exist for
+   * the requested type, and calls the specification validator, if it exists, on
+   * the constructed caveat.
+   *
+   * Throws an error if validation fails.
+   *
+   * @param origin - The origin of the permission subject.
+   * @param target - The permission target name.
+   * @param requestedCaveat - The requested caveat to construct.
+   * @returns The constructed caveat.
+   */
+  private constructCaveat(
+    origin: OriginString,
+    target: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    requestedCaveat: unknown,
+  ): ExtractCaveats<ControllerCaveatSpecification> {
+    if (!isPlainObject(requestedCaveat)) {
+      throw new InvalidCaveatError(requestedCaveat, origin, target);
+    }
+
+    if (Object.keys(requestedCaveat).length !== 2) {
+      throw new InvalidCaveatFieldsError(requestedCaveat, origin, target);
+    }
+
+    const { type, value } = requestedCaveat;
+    if (typeof type !== 'string') {
+      throw new InvalidCaveatTypeError(requestedCaveat, origin, target);
+    }
+
+    if (!hasProperty(requestedCaveat, 'value')) {
+      throw new CaveatMissingValueError(requestedCaveat, origin, target);
+    }
+
+    // Assert that the caveat value is valid JSON
+    if (!deepEqual(value, JSON.parse(JSON.stringify(value)))) {
+      throw new CaveatInvalidJsonError(requestedCaveat, origin, target);
+    }
+
+    const caveat = _constructCaveat(type, (value as Json) ?? null);
```

Comment:
Nit: we could probably get rid of this `as` by moving the JSON validation to a utility function that uses a type predicate

Reply:
11e318e

Reply:
Looks great! But you forgot the last step of removing the cast

```suggestion
    const caveat = _constructCaveat(type, value ?? null);
```